internal/gmm/sm.go

```diff
@@ -204,6 +205,15 @@ func Authentication(state *fsm.State, event fsm.EventType, args fsm.ArgsType) {
 	case AuthFailEvent:
 		logger.GmmLog.Debugln(event)
 		logger.GmmLog.Warnln("Reject authentication")
+		amfUe := args[ArgAmfUe].(*context.AmfUe)
+		accessType := args[ArgAccessType].(models.AccessType)
+		if amfUe.RanUe[accessType] != nil {
+			err := amfUe.RanUe[accessType].Remove()
```

Comment:
should send UEContextReleaseCommand to RAN before RanUe removed.

Reply:
Added `SendUEContextReleaseCommand()` before RanUe removed.

Reply:
RanUe need to wait for UEContextReleaseComplete.
It's better:
```
ranUe := amfUe.RanUe[accessType]
if ranUe != nil {
	ngap_message.SendUEContextReleaseCommand(ranUe, context.UeContextN2NormalRelease,
		ngapType.CausePresentNas, ngapType.CauseNasPresentAuthenticationFailure)

	// keep RanUe to wait for UEContextReleaseComplete
	ranUe.DetachAmfUe()
	amfUe.DetachRanUe(accessType)
	gmm_common.RemoveAmfUe(amfUe)
```

Reply:
It is better to remove the RanUe first, since the UEContextReleaseRequest (gNB initiated) may come or the UE may reconnect before the UEContextReleaseComplete is received.

Reply:
another way:
```
ranUe := amfUe.RanUe[accessType]
if ranUe != nil {
	ngap_message.SendUEContextReleaseCommand(ranUe, context.UeContextN2NormalRelease,
		ngapType.CausePresentNas, ngapType.CauseNasPresentAuthenticationFailure)
	// keep ranUe to wait for UEContextReleaseComplete
	ranUe.DetachAmfUe()
	amfUe.DetachRanUe(accessType)
	gmm_common.RemoveAmfUe(amfUe)
}
```

Reply:
If AMF leaves RanUe and UEContextReleaseRequest (gNB initiated) comes in immediately after, doesn't that mean sending double UEContextReleaseCommand?

Reply:
If AMF receives UEContextReleaseRequest after sending UEContextReleaseCommand, AMF should skip the UEContextReleaseRequest.

Reply:
In that case, the following UEContextReleaseCommand sending process would be executed without skipping.
https://github.com/free5gc/amf/blob/e839de03774f7c5b06dbaaa368ea6f0813a4d0d9/internal/ngap/handler.go#L1963

Reply:
so you mean if Removing RanUe, AMF will ignore UEContextReleaseRequest or UEContextReleaseComplete when receiving them?

Reply:
AMF returns ErrorIndication, but it is not particularly harmful, so there is no problem.